src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
 			break;
 			}
 
-	// Ideally, we would always call select on the fds to see which
-	// are ready, and return the soonest. Unfortunately, that'd mean
-	// one select-call per packet, which we can't afford in high-volume
-	// environments.  Thus, we call select only every SELECT_FREQUENCY
-	// call (or if all sources report that they are dry).
+	// If there aren't any sources and exit_only_after_terminate is false, just
+	// return an empty set of sources. We want the main loop to end.
+	if ( Size() == 0 && ( ! BifConst::exit_only_after_terminate || terminating ) )
+		return ready;
 
-	++call_count;
+	double timeout = -1;
+	IOSource* timeout_src = nullptr;
+	timer_expired = false;
 
-	IOSource* soonest_src = 0;
-	double soonest_ts = 1e20;
-	double soonest_local_network_time = 1e20;
-	bool all_idle = true;
+	PktSrc* pkt_src = pkt_srcs.empty() ? nullptr : pkt_srcs.front();
 
-	// Find soonest source of those which tell us they have something to
-	// process.
-	for ( SourceList::iterator i = sources.begin(); i != sources.end(); ++i )
+	// If we have an active source of data then find the next timeout value from
+	// those sources. Start by getting the next timeout from the timer manager
+	// since it's not an IOSource.
+	// TODO: how do we handle multiple timer managers here?
+	if ( pkt_src || broker_mgr->Active() )
```

Comment:
Don't understand this condition: what's special about pkt-sources and Broker that we go into this only if we have one of those; why not generally return use next timeout (and I guess potentially just sleep)? Or said differently: would be nice if we didn't have to hardcode especially Broker here.

Reply:
If one of those isn't true, then the loop just spins at 100ms waits. Specifically, the `broker.store.ops` test fails because we set a timer for 3.x seconds. Without spinning in that case, nothing is forcing events to happen (because `mgr.Drain()` is called in `net_run`) and so the rest of the stuff that's supposed to happen in that test never takes place. Eventually the timeout happens and everything in the test times out.

This may honestly point at something else systemic with the way that I did this though, where it breaks if something is relying on events to be processed and we're waiting on a timer.

Reply:
I've been thinking over this one this weekend. What do you think about making the event manager an IOSource, and letting it handle events as during the loop like everything else? I haven't really dug into how the event manager works. Does it depend too heavily on what packets got processed earlier during the loop to know what it needs to do?

Reply:
It's primarily a timing thing I think, some pieces may expect events to have been processed already. It also always lets all queued events fire at once; we'd need to keep that. There are also a couple of  places that drain the event mgr explicitly to force processing "right now". I'm guessing that as long as we keep that ability (i.e., have a way to drain events outside of the normal IO loop processing), it should work. 

Would be easy to hack this change in and run tests to see if something fundamentally breaks?

Meta comment, as this is coming up at a few places: The more we can unify things into a single I/O source API, the better. So if we could make the timer mgrs and the event mgr all IOSources, and also avoid special-casing specific I/O sources somewhere (packet sources, Broker, etc.), that would be quite neat!

Reply:
Only one additional test fails with that hacked in: `core.recursive-event`. Strangely, `broker.store.ops` succeeds if I remove the `-b` from the command line used in the test, but I haven't investigated why. For `core.recursive-event`, it's supposed to print 9 different timestamps, but instead only prints 1.